pkg/summarizer/flakiness.go

```diff
@@ -0,0 +1,286 @@
+
+	for _, test := range grid.Rows {
+		// each row in Rows will have a field results
+		// results is run length encoded
+		// e.g. (encoded -> decoded equivalent):
+		// [0, 3, 5, 4] -> [0, 0, 0, 5, 5, 5, 5]
+		// decoded int values correspond to Row.Result enum
+		var resultCounts Result
+		resultCounts.infraFailures = make(map[string]int)
+		i := -1
+		for testResult := range result.Iter(ctx, test.Results) {
+			i += 1
+			if !isWithinTimeFrame(grid.Columns[i], startTime, endTime) {
+				continue
+			}
+			switch rowResult := result.Coalesce(testResult, result.IgnoreRunning); rowResult {
+			case state.Row_NO_RESULT:
+				continue
+			case state.Row_FAIL:
+				categorizeFailure(&resultCounts, test.Messages[i])
+			case state.Row_PASS:
+				resultCounts.passed += 1
+			case state.Row_FLAKY:
+				getValueOfFlakyResult(&resultCounts)
+			}
+		}
+		if resultCounts.failed > 0 || resultCounts.passed > 0 || resultCounts.flakyCount > 0 {
+			resultCounts.name = test.Name
+			results = append(results, resultCounts)
+		}
+
+	}
+	return results
+}
+
+func analyzeFlakinessFromResults(results []Result, startTime int, endTime int, tab string) Healthiness {
+	return naiveFlakiness(results, MIN_RUNS, startTime, endTime, tab)
+}
+
+func naiveFlakiness(results []Result, minRuns int, startDate int, endDate int, tab string) Healthiness {
+	testByEnv := make(map[string]TestInfo)
+	infraIssues := make(map[string]int)
+
+	for _, test := range results {
+		name, env := getNameAndEnvFromTest(test.name, tab)
+		if len(test.infraFailures) > 0 {
+			for errorType, errorCount := range test.infraFailures {
+				infraIssues[test.name+"-"+errorType] += errorCount
+			}
+		}
+
+		testInfo, success := calculateNaiveFlakiness(test, minRuns)
+		if !success {
+			continue
+		}
+		testInfo.name = name
+		testInfo.env = env
+
+		if currTestInfo, exists := testByEnv[name]; !exists || currTestInfo.flakiness < testInfo.flakiness {
+			testByEnv[name] = testInfo
+		}
+	}
+	// Populate Healthiness with above calculated information
+	healthiness := createHealthiness(startDate, endDate, results, testByEnv, infraIssues)
+	return healthiness
+}
+
+func getNameAndEnvFromTest(testName string, tabName string) (string, string) {
+	name := testName
+	env := tabName
+	match := JAILED_REGEX.FindStringSubmatch(name)
+	if match != nil {
+		name = match[2]
+		env = match[3]
+	} else {
+		logrus.Infof("Test \"%s\" could not be split into name and env, using tab name \"%s\" as its env", testName, tabName)
+	}
+	return name, env
+}
+
+func calculateNaiveFlakiness(test Result, minRuns int) (TestInfo, bool) {
+	failedCount := test.failed
+	totalCount := test.passed + test.failed
+	totalCountWithInfra := totalCount + test.failedInfraCount
+	if totalCount < minRuns {
+		return TestInfo{}, false
+	}
+	flakiness := 100 * float64(failedCount) / float64(totalCount)
+	infraInfo := calculateInfraInfo(test.infraFailures, test.failedInfraCount)
+	testInfo := TestInfo{
+		name:               "",
```

Comment:
If they are going to be the default value, do you need to add them?

Reply:
Done